apps/passport-server/src/services/generic-issuance/subservices/PipelineSubservice.ts

```diff
@@ -265,6 +266,41 @@ export class PipelineSubservice {
   ): Promise<UpsertPipelineResult> {
     return traced(SERVICE_NAME, "upsertPipelineDefinition", async () => {
       logger(SERVICE_NAME, "upsertPipelineDefinition", str(newDefinition));
+
+      const ids = uniqueIdsForPipelineDefinition(newDefinition);
+      const seen = new Set<string>();
+      for (const id of ids) {
+        if (seen.has(id)) {
+          throw new PCDHTTPError(
+            401,
+            `ID ${id} is used more than once in this pipeline`
+          );
+        }
+        seen.add(id);
+      }
+
+      const otherPipelines = (
+        await this.pipelineDB.loadPipelineDefinitions()
+      ).filter((definition) => definition.id !== newDefinition.id);
+
+      for (const definition of otherPipelines) {
+        const otherPipelineIds = new Set(
+          uniqueIdsForPipelineDefinition(definition).values()
+        );
+
+        for (const id of ids) {
+          if (otherPipelineIds.has(id)) {
+            const otherPipelineName = definition.options.name
+              ? `${definition.options.name} (${definition.id})`
+              : `pipeline ${definition.id}`;
+            throw new PCDHTTPError(
+              401,
+              `ID ${id} is already in use by ${otherPipelineName} and cannot be used by this pipeline`
+            );
+          }
+        }
+      }
+
```

Comment:
suggest moving this logic into the standalone `upsertPipelineDefinition` function (which has its own file) just for consistency - my intention is that the function `upsertPipelineDefinition` in `PipelineSubservice` is basically a proxy for that function.

Reply:
Done